### ISSUE TITLE
Add retention period for dead letter queue

### ISSUE THREAD

**Issue by suyograo** [elastic/logstash#6990]
Add a feature to control the retention period for the dead letter queue. Retention can be time-based or age-based. After the threshold has reached, data should be deleted from the dead letter starting from the earliest file.

```
dead_letter_queue.size.threshold:  4g
dead_letter_queue.age.threshold:  24h
```

**Comment by robbavey** [comment]
Couple of general questions/assumptions:

- Is it fair to make the assumption that retention is performed at the segment level?
- Are these values global, or per-pipeline? This has particular implications for the size threshold...
- Is the size threshold a hard limit (queue cannot go above the threshold), or one that advises that old segments will be cleaned up once the queue size breaches the threshold?

**Comment by suyograo** [comment]
> Is it fair to make the assumption that retention is performed at the segment level?

Yes.

> Are these values global, or per-pipeline? This has particular implications for the size threshold...

IMO, we should make it a global setting. This would keep the settings simple. It is not clear why users would want a different threshold for a different pipeline.

> Is the size threshold a hard limit (queue cannot go above the threshold), or one that advises that old segments will be cleaned up once the queue size breaches the threshold?

We should start deleting old content once the threshold is reached. This basically ensures that DLQ always has space to accept newer events within a finite disk space. If the LS admin never cleans the DLQ events, it means those events are not important to them and can be wiped anyway. This is similar to how Kafka's retention works.

**Comment by robbavey** [comment]
> IMO, we should make it a global setting. This would keep the settings simple. It is not clear why users would want a different threshold for a different pipeline.

Agreed that we wouldn't necessarily want a different threshold, and would prefer a global threshold. My question was more whether the threshold should be applied per dlq instance (ie, per pipeline), or across all instances together. Per dlq instance would be simpler, but we should make a conscious decision.

> This is similar to how Kafka's retention works.

This was the model I was thinking of ;)

**Comment by suyograo** [comment]
Yeah, applying the retention per DLQ makes sense vs all. You want the retention to be independent of DLQs used in other pipelines. I've seen this work well in systems like Kafka.

**Comment by suyograo** [comment]
^^^ this would enable us to expose retention settings per pipeline in the future if users ask for it.

**Comment by schmandforke** [comment]
+1

**Comment by kjakman** [comment]
There might still be a few old zombies in the DLQ, I hear. Is this the way, or plan B?

**Comment by vivekkarne** [comment]
+1

**Comment by arpit-ghiya** [comment]
+1
The document should at least talk about how to clean up these files manually. Do we need to stop logstash or just `rm -rf` ?

**Comment by QCU266** [comment]
+1000

**Comment by loakley-he** [comment]
Seriously this would be great still.

**Comment by eugene-marchanka** [comment]
looking forward for this improvement 🙏🏻

**Comment by roaksoax** [comment]
Logstash 8.4 now features an [Age Retention Policy ](https://www.elastic.co/guide/en/logstash/master/dead-letter-queues.html#age-policy). As such, we believe this issue is now resolve.

**Comment by eugene-marchanka** [comment]
I see this in `master` branch but it is not released yet.
Is this functionality already working?

**Comment by yaauie** [comment]
@eugene-marchanka I see it [in 8.5](https://www.elastic.co/guide/en/logstash/8.5/dead-letter-queues.html#age-policy), the release of which is in-flight right now and will be finalized in the next week or so. Comments above indicate 8.4 has it, but I _think_ it may have been merged to `main` _before_ the 8.4.0 release but _after_ the 8.4.0 feature-freeze.
